cli/src/commands/extensions/api.rs

```diff
@@ -381,69 +443,124 @@ async fn parse_lockfile(
 #[op]
 #[cfg(unix)]
 fn run_sandboxed(process: Process) -> Result<ProcessOutput> {
-    // Setup process to be run.
-    let mut command = Command::new(&process.cmd);
-    command.args(&process.args);
-    command.stdin(process.stdin);
-    command.stdout(process.stdout);
-    command.stderr(process.stderr);
-
-    // Apply sandbox to subprocess after fork.
-    unsafe {
-        command.pre_exec(move || {
-            fn into_ioerr<E: Into<Box<dyn std::error::Error + Send + Sync>>>(err: E) -> io::Error {
-                io::Error::new(io::ErrorKind::Other, err)
-            }
-
-            let home_dir = dirs::home_dir().map_err(into_ioerr)?;
-
-            let mut birdcage = if process.exceptions.strict {
-                Birdcage::new().map_err(into_ioerr)?
+    let mut stdout_fds: ProcessStdioFds = process.stdout.try_into()?;
+    let mut stderr_fds: ProcessStdioFds = process.stderr.try_into()?;
+
+    match unsafe { libc::fork() } {
+        -1 => Err(io::Error::last_os_error().into()),
+        // Handle child process.
+        0 => {
+            // Connect STDOUT/STDERR with parent if necessary.
+            stdout_fds.replace_fd(libc::STDOUT_FILENO)?;
+            stderr_fds.replace_fd(libc::STDERR_FILENO)?;
+
+            // Apply sandboxing rules.
+            lock_process(process.exceptions)?;
```

Comment:
The issue mainly seems to focus around spawning children, but we never actually do any of that in our code. The fork only allocates to setup the sandboxing stuff, but the actual child is behind another fork that should be secured by Rust's internal command structure?

I'll admit that I'm not intimately familiar with the issue itself, but I wouldn't be surprised if our usage was not a problem at all.

I'm personally probably more concerned about this (see `man 2 fork`):
> After  a  fork() in a multithreaded program, the child can safely call only async-signal-safe functions (see signal-safety(7)) until such time as it calls execve(2).

Reply:
> The issue mainly seems to focus around spawning children, but we never actually do any of that in our code.

I'm not sure I understand... We call `libc::fork()` which spawns a child process

> I'm personally probably more concerned about this (see man 2 fork):

I think this is the same issue. Functions that are async-signal-safe must not rely on any global mutexes, etc. And `malloc(3)` and `mmap(2)` are not in the list of async-signal-safe functions because they rely on global mutexes.

Reply:
> I'm not sure I understand... We call libc::fork() which spawns a child process

I guess I wasn't precise enough here. The Rust comment seems very specific about `do_exec` and explicitly mentions `spawn`. Not necessarily about `fork`?

> I think this is the same issue. Functions that are async-signal-safe must not rely on any global mutexes, etc. And malloc(3) and mmap(2) are not in the list of async-signal-safe functions because they rely on global mutexes.

I'm not sure it is, considering the Rust code only mentions freebsd/macos having problems with malloc, while the manpage definitely implies that Linux is affected too should signals get involved. Now to be fair the amount of C code that isn't properly signal safe and still works just fine is probably higher than the amount of signal safe C code, but that doesn't mean we shouldn't at least try.